### Mistral inference

In [ ]:
from huggingface_hub.hf_api import HfFolder
HfFolder.save_token('')

In [2]:
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch


class Server:

    def __init__(self):
        # model_id = '/kaggle/input/mistral/pytorch/7b-instruct-v0.1-hf/1'
        model_id = "mistralai/Mistral-7B-Instruct-v0.3"
        self.tokenizer = AutoTokenizer.from_pretrained(model_id)
        self.model = AutoModelForCausalLM.from_pretrained(
            model_id, torch_dtype=torch.bfloat16, device_map="cuda")
        self.chat_context = []

    def add_system_context(self, message):
        if len(self.chat_context) == 0:
            self.chat_context.insert(0, {"role": "system", "content": message})
        else:
            if self.chat_context[0]["role"] == "system":
                self.chat_context[0]["content"] = message
            else:
                self.chat_context.insert(
                    0, {"role": "system", "content": message})

    def add_user_context(self, message):
        self.chat_context.append({"role": "user", "content": message})

    def add_model_context(self, message):
        self.chat_context.append({"role": "assistant", "content": message})

    def ask_question(self, message, add_context=False):
        inputs = self.tokenizer.apply_chat_template(
            self.chat_context + [{"role": "user", "content": message}],
            add_generation_prompt=True,
            tokenize=False)

        inputs = self.tokenizer([inputs], return_tensors="pt")
        inputs.to(self.model.device)

        generated_ids = self.model.generate(
            inputs.input_ids, max_new_tokens=1000)

        generated_ids = [
            output_ids[len(input_ids):] for input_ids, output_ids in zip(inputs.input_ids, generated_ids)
        ]

        response = self.tokenizer.batch_decode(
            generated_ids, skip_special_tokens=True)[0]

        if add_context:
            self.add_user_context(message)
            self.add_model_context(response.text)

        return response

In [3]:
mistral = Server()
mistral.add_system_context("Answer as Megatron")
print(mistral.ask_question("What is Linux?"))
print(mistral.ask_question("What is asked in previous question?"))

tokenizer_config.json:   0%|          | 0.00/141k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/587k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.96M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/414 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/601 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/23.9k [00:00<?, ?B/s]

model-00001-of-00003.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

model-00002-of-00003.safetensors:   0%|          | 0.00/5.00G [00:00<?, ?B/s]

model-00003-of-00003.safetensors:   0%|          | 0.00/4.55G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.


Greetings, human. Linux is an open-source operating system based on the Unix operating system. It was first created by Linus Torvalds in 1991. Unlike proprietary operating systems, Linux is free to use, modify, and distribute. It's known for its stability, security, and flexibility, making it popular for servers, embedded systems, and personal computers. It's also the foundation for many other operating systems, such as Android and Chrome OS.
The previous question asked for a response in the voice of Megatron, a character from the Transformers franchise. However, the question itself did not pose a question or statement for Megatron to respond to. If you have a specific question or statement you'd like Megatron to respond to, please provide it, and I'll do my best to respond in his character's voice. For example, you could ask, "Megatron, what is your ultimate goal?" or "Megatron, how do you feel about the Autobots?" and I'll provide a response in character.


## Ontology access

In [3]:
!pip install owlready2
!pip install rdflib

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.3/27.3 MB 78.1 MB/s eta 0:00:0000:0100:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for owlready2: filename=owlready2-0.47-cp310-cp310-linux_x86_64.whl size=23515870 sha256=6b06bb82337590f4b5d66594e910079486416c0d776a7e178e3920326b7f1a20
  Stored in directory: /root/.cache/pip/wheels/27/3e/ba/4171c4b10bba9fe1774fbf8fcf794de889e636ce64ad83a533
Successfully built owlready2
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 562.4/562.4 kB 11.9 MB/s eta 0:00:0000:01


In [4]:
from owlready2 import *
import rdflib
import logging
from pathlib import Path
import pandas as pd
from enum import Enum

In [5]:
class AnnotationURIs(object):
    '''
    This class manages the most common ontology annotations
    '''

    def __init__(self):       
        self.mainLabelURIs = set()
        self.synonymLabelURIs = set()
        self.lexicalAnnotationURIs = set()      
                        
        #Main labels
        self.mainLabelURIs.add("http://www.w3.org/2000/01/rdf-schema#label")
        self.mainLabelURIs.add("http://www.w3.org/2004/02/skos/core#prefLabel")
        self.mainLabelURIs.add("http://purl.obolibrary.org/obo/IAO_0000111")
        self.mainLabelURIs.add("http://purl.obolibrary.org/obo/IAO_0000589")

        #synonyms or alternative names
        self.synonymLabelURIs.add("http://www.geneontology.org/formats/oboInOwl#hasRelatedSynonym")
        self.synonymLabelURIs.add("http://www.geneontology.org/formats/oboInOwl#hasExactSynonym")
        self.synonymLabelURIs.add("http://www.geneontology.org/formats/oboInOWL#hasExactSynonym")
        self.synonymLabelURIs.add("http://www.geneontology.org/formats/oboInOwl#hasRelatedSynonym")
        self.synonymLabelURIs.add("http://purl.bioontology.org/ontology/SYN#synonym")
        self.synonymLabelURIs.add("http://scai.fraunhofer.de/CSEO#Synonym")
        self.synonymLabelURIs.add("http://purl.obolibrary.org/obo/synonym")
        self.synonymLabelURIs.add("http://ncicb.nci.nih.gov/xml/owl/EVS/Thesaurus.owl#FULL_SYN")
        self.synonymLabelURIs.add("http://www.ebi.ac.uk/efo/alternative_term")
        self.synonymLabelURIs.add("http://ncicb.nci.nih.gov/xml/owl/EVS/Thesaurus.owl#Synonym")
        self.synonymLabelURIs.add("http://bioontology.org/projects/ontologies/fma/fmaOwlDlComponent_2_0#Synonym")
        self.synonymLabelURIs.add("http://www.geneontology.org/formats/oboInOwl#hasDefinition")
        self.synonymLabelURIs.add("http://bioontology.org/projects/ontologies/birnlex#preferred_label")
        self.synonymLabelURIs.add("http://bioontology.org/projects/ontologies/birnlex#synonyms")
        self.synonymLabelURIs.add("http://www.w3.org/2004/02/skos/core#altLabel")
        self.synonymLabelURIs.add("https://cfpub.epa.gov/ecotox#latinName")
        self.synonymLabelURIs.add("https://cfpub.epa.gov/ecotox#commonName")
        self.synonymLabelURIs.add("https://www.ncbi.nlm.nih.gov/taxonomy#scientific_name")
        self.synonymLabelURIs.add("https://www.ncbi.nlm.nih.gov/taxonomy#synonym")
        self.synonymLabelURIs.add("https://www.ncbi.nlm.nih.gov/taxonomy#equivalent_name")
        self.synonymLabelURIs.add("https://www.ncbi.nlm.nih.gov/taxonomy#genbank_synonym")
        self.synonymLabelURIs.add("https://www.ncbi.nlm.nih.gov/taxonomy#common_name")       

        #Alternative term       
        self.synonymLabelURIs.add("http://purl.obolibrary.org/obo/IAO_0000118")
        #Mouse anatomy
        #Lexically rich interesting
        self.lexicalAnnotationURIs.update(self.mainLabelURIs)
        self.lexicalAnnotationURIs.update(self.synonymLabelURIs)
        self.lexicalAnnotationURIs.add("http://www.w3.org/2000/01/rdf-schema#comment")
        self.lexicalAnnotationURIs.add("http://www.geneontology.org/formats/oboInOwl#hasDbXref")
        self.lexicalAnnotationURIs.add("http://purl.org/dc/elements/1.1/description")
        self.lexicalAnnotationURIs.add("http://purl.org/dc/terms/description")
        self.lexicalAnnotationURIs.add("http://purl.org/dc/elements/1.1/title")
        self.lexicalAnnotationURIs.add("http://purl.org/dc/terms/title")

        #Definition
        self.lexicalAnnotationURIs.add("http://purl.obolibrary.org/obo/IAO_0000115")
        #Elucidation
        self.lexicalAnnotationURIs.add("http://purl.obolibrary.org/obo/IAO_0000600")
        #has associated axiomm fol
        self.lexicalAnnotationURIs.add("http://purl.obolibrary.org/obo/IAO_0000602")
        #has associated axiomm nl
        self.lexicalAnnotationURIs.add("http://purl.obolibrary.org/obo/IAO_0000601")
        self.lexicalAnnotationURIs.add("http://www.geneontology.org/formats/oboInOwl#hasOBONamespace")

    def getAnnotationURIsForPreferredLabels(self):
        return self.mainLabelURIs
    
    def getAnnotationURIsForSymnonyms(self):
        return self.synonymLabelURIs

    def getAnnotationURIsForLexicalAnnotations(self):
        return self.lexicalAnnotationURIs


In [6]:
class Reasoner(Enum):
    HERMIT=0 #Not really adding the right set of entailments
    PELLET=1 #Slow for large ontologies
    STRUCTURAL=2  #Basic domain/range propagation
    NONE=3 #No reasoning

class OntologyAccess(object):

    def __init__(self, urionto, annotate_on_init=True):
        logging.basicConfig(format='%(levelname)s: %(message)s', level=logging.DEBUG)
        self.urionto = str(urionto)
        if annotate_on_init:
            self.loadOntology()
            self.indexAnnotations()

    def getOntologyIRI(self):
        return self.urionto

    def loadOntology(self, reasoner=Reasoner.NONE, memory_java='10240'):
        #Method from owlready
        self.onto = get_ontology(self.urionto).load()
        owlready2.reasoning.JAVA_MEMORY=memory_java
        owlready2.set_log_level(9)

        if reasoner==Reasoner.PELLET:
            try:
                with self.onto:  #it does add inferences to ontology

                    # Is this wrt data assertions? Check if necessary
                    # infer_property_values = True, infer_data_property_values = True
                    logging.info("Classifying ontology with Pellet...")
                    sync_reasoner_pellet() #it does add inferences to ontology
                    unsat = len(list(self.onto.inconsistent_classes()))
                    logging.info("Ontology successfully classified.")
                    if unsat > 0:
                        logging.warning("There are " + str(unsat) + " unsatisfiabiable classes.")
            except:
                logging.info("Classifying with Pellet failed.")

        elif reasoner==Reasoner.HERMIT:
                try:
                    with self.onto:  #it does add inferences to ontology
                        logging.info("Classifying ontology with HermiT...")
                        sync_reasoner() #HermiT doe snot work very well....
                        unsat = len(list(self.onto.inconsistent_classes()))
                        logging.info("Ontology successfully classified.")
                        if unsat > 0:
                            logging.warning("There are " + str(unsat) + " unsatisfiabiable classes.")
                except:
                    logging.info("Classifying with HermiT failed.")

        self.graph = default_world.as_rdflib_graph()
        logging.info("There are {} triples in the ontology".format(len(self.graph)))
        #self.graph = self.world.as_rdflib_graph()

    def getOntology(self):
        return self.onto

    def getClassByURI(self, uri):
        for cls in list(self.getOntology().classes()):
            if (cls.iri==uri):
                return cls
        return None

    def getClassByName(self, name):
        for cls in list(self.getOntology().classes()):
            if (cls.name.lower()==name.lower()):
                return cls
        return None

    def getEntityByURI(self, uri):
        for cls in list(self.getOntology().classes()):
            if (cls.iri==uri):
                return cls
        for prop in list(self.getOntology().properties()):
            if (prop.iri==uri):
                return prop
        return None

    def getEntityByName(self, name):
        for cls in list(self.getOntology().classes()):
            if (cls.name.lower()==name.lower()):
                return cls
        for prop in list(self.getOntology().properties()):
            if (prop.name.lower()==name.lower()):
                return prop
        return None

    def getClassObjectsContainingName(self, name):
        classes = []
        for cls in list(self.getOntology().classes()):
            if (name.lower() in cls.name.lower()):
                classes.append(cls)
        return classes

    def getClassIRIsContainingName(self, name):
        classes = []
        for cls in list(self.getOntology().classes()):
            if (name.lower() in cls.name.lower()):
                classes.append(cls.iri)
        return classes

    def getAncestorsURIsMinusClass(self,cls):
        ancestors_str = self.getAncestorsURIs(cls)
        ancestors_str.remove(cls.iri)
        return ancestors_str

    def getAncestorsURIs(self,cls):
        ancestors_str = set()
        for anc_cls in cls.ancestors():
            ancestors_str.add(anc_cls.iri)
        return ancestors_str

    def getAncestorsNames(self,cls):
        ancestors_str = set()
        for anc_cls in cls.ancestors():
            ancestors_str.add(anc_cls.name)
        return ancestors_str

    def getAncestors(self,cls):
        ancestors_str = set()
        for anc_cls in cls.ancestors():
            ancestors_str.add(anc_cls)
        return ancestors_str

    def getDescendantURIs(self,cls):
        descendants_str = set()
        for desc_cls in cls.descendants():
            descendants_str.add(desc_cls.iri)
        return descendants_str

    def getDescendantNames(self,cls):
        descendants_str = set()
        for desc_cls in cls.descendants():
            descendants_str.add(desc_cls.name)
        return descendants_str

    def getDescendants(self,cls):
        descendants_str = set()
        for desc_cls in cls.descendants():
            descendants_str.add(desc_cls)
        return descendants_str

    def getDescendantNamesForClassName(self, cls_name):
        cls = self.getClassByName(cls_name)
        descendants_str = set()
        for desc_cls in cls.descendants():
            descendants_str.add(desc_cls.name)
        return descendants_str

    def isSubClassOf(self, sub_cls1, sup_cls2):
        if sup_cls2 in sub_cls1.ancestors():
            return True
        return False

    def isSuperClassOf(self, sup_cls1, sub_cls2):
        if sup_cls1 in sub_cls2.ancestors():
            return True
        return False

    def getDomainURIs(self, prop):
        domain_uris = set()
        for cls in prop.domain:
            try:
                domain_uris.add(cls.iri)
            except AttributeError:
                pass
        return domain_uris

    def getDatatypeRangeNames(self, prop):
        range_uris = set()
        for cls in prop.range:
            range_uris.add(cls.name)  #datatypes are returned without uri
        return range_uris

    #Only for object properties
    def getRangeURIs(self, prop):
        range_uris = set()
        for cls in prop.range:
            try:
                range_uris.add(cls.iri)
            except AttributeError:
                pass
        return range_uris

    def getInverses(self, prop):
        inv_uris = set()
        for p in prop.inverse:
            inv_uris.add(p.iri)
        return inv_uris

    def getClasses(self):
        return self.getOntology().classes()

    def getDataProperties(self):
        return self.getOntology().data_properties()

    def getObjectProperties(self):
        return self.getOntology().object_properties()

    def getIndividuals(self):
        return self.getOntology().individuals()

    def getGraph(self):
        return self.graph

    def queryGraph(self, query):
        results = self.graph.query(query)
        return list(results)


    def getQueryForAnnotations(self, ann_prop_uri):
        return """SELECT DISTINCT ?s ?o WHERE {{
        {{
        ?s <{ann_prop}> ?o .
        }}
        UNION
        {{
        ?s <{ann_prop}> ?i .
        ?i <http://www.w3.org/2000/01/rdf-schema#label> ?o .
        }}
        }}""".format(ann_prop=ann_prop_uri)


    def indexAnnotations(self):
        annotation_uris = AnnotationURIs()
        self.entityToSynonyms = {}
        self.allEntityAnnotations = {}
        self.preferredLabels = {}
        self.populateAnnotationDicts(annotation_uris.getAnnotationURIsForSymnonyms(), self.entityToSynonyms)
        self.populateAnnotationDicts(annotation_uris.getAnnotationURIsForLexicalAnnotations(), self.allEntityAnnotations)
        self.populateAnnotationDicts(annotation_uris.getAnnotationURIsForPreferredLabels(), self.preferredLabels)

    def populateAnnotationDicts(self, annotation_uris, dictionary):
        for ann_prop_uri in annotation_uris:
            results = self.queryGraph(self.getQueryForAnnotations(ann_prop_uri))
            for row in results:
                try:
                    if row[1].language=="en" or row[1].language==None:
                        if not str(row[0]) in dictionary:
                            dictionary[str(row[0])]=set()
                        dictionary[str(row[0])].add(row[1].value)
                except AttributeError:
                    pass
        return None

    def getSynonymsNames(self, entity):
        if entity.iri not in self.entityToSynonyms: 
            return {}
        return self.entityToSynonyms[entity.iri]

    def getAnnotationNames(self, entity):
        if entity.iri not in self.allEntityAnnotations: 
            return {}
        return self.allEntityAnnotations[entity.iri]

    def getPrefferedLabels(self, entity):
        if entity.iri not in self.preferredLabels: 
            return {}
        return self.preferredLabels[entity.iri]


In [7]:
# check if the enviroment is kaggle or local
DATASET_PATH = Path('/kaggle/input/')
if not DATASET_PATH.exists():
    DATASET_PATH = Path('.')
DATASET_PATH /= 'rai-bioml-2024'

src_onto_path = DATASET_PATH / 'omim-ordo/omim.owl'
tgt_onto_path = DATASET_PATH / 'omim-ordo/ordo.owl'

onto_src = OntologyAccess(src_onto_path, annotate_on_init=False)
onto_src.loadOntology()
onto_src.indexAnnotations()

onto_tgt = OntologyAccess(tgt_onto_path, annotate_on_init=True)

* Owlready2 * Creating new ontology ordo </kaggle/input/rai-bioml-2024/omim-ordo/ordo.owl#>.
* Owlready2 * ADD TRIPLE /kaggle/input/rai-bioml-2024/omim-ordo/ordo.owl http://www.w3.org/1999/02/22-rdf-syntax-ns#type http://www.w3.org/2002/07/owl#Ontology
* Owlready2 *     ...loading ontology ordo from /kaggle/input/rai-bioml-2024/omim-ordo/ordo.owl...


* Owlready2 * Reseting property ann.use_in_alignment: new triples are now available.


* Owlready2 *     ...41 properties found: BFO_0000050, Orphanet_317343, Orphanet_317344, Orphanet_317345, Orphanet_317346, Orphanet_317348, Orphanet_317349, Orphanet_327767, Orphanet_410295, Orphanet_410296, Orphanet_465410, Orphanet_C016, Orphanet_C017, Orphanet_C020, Orphanet_C022, Orphanet_C025, Orphanet_C026, Orphanet_C027, Orphanet_C024, Orphanet_C028, Orphanet_C029, Orphanet_C030, Orphanet_C032, Orphanet_C040, label, versionInfo, use_in_alignment, alternative_term, ECO_0000205, ECO_0000218, creator, created, license, modified, definition, definition_citation, reason_for_obsolescence, symbol, notation, permits, requires


In [8]:
onto_classes = [x for x in onto_src.getClasses()]
print(onto_classes[:25], type(onto_classes[0]))

[entry.100050, entry.100070, phenotypicSeries.PS100070, entry.100100, entry.100200, entry.100300, phenotypicSeries.PS100300, entry.100600, entry.100700, entry.100800, entry.100820, entry.101000, entry.101120, entry.101200, entry.101400, entry.101600, entry.101800, phenotypicSeries.PS101800, entry.101805, entry.101840, entry.101850, entry.101900, entry.102000, entry.102100, entry.102150] <class 'owlready2.entity.ThingClass'>


In [9]:
ancestors = {x:values for x in onto_src.getClasses() if len(values:=onto_src.getAncestors(x))>3}
ancestors_labels = {'\n'.join(onto_src.getPrefferedLabels(x)):[
    onto_src.getPrefferedLabels(y) for y in values] 
                    for x, values in ancestors.items()}
dict(list(ancestors_labels.items())[:5])

{'frontotemporal dementia and/or amyotrophic lateral sclerosis 1': [{'Frontotemporal dementia and/or amyotrophic lateral sclerosis'},
  {'frontotemporal dementia and/or amyotrophic lateral sclerosis 1'},
  {},
  {'Amyotrophic lateral sclerosis'}],
 'albinism, oculocutaneous,  iia 6': [{'albinism, oculocutaneous,  iia 6'},
  {'Oculocutaneous albinism'},
  {'Skin/hair/eye pigmentation, variation in'},
  {}],
 'episodic kinesigenic dyskinesia 1': [{'Dystonia'},
  {'episodic kinesigenic dyskinesia 1'},
  {},
  {'Episodic kinesigenic dyskinesia'}],
 'coffin-siris syndrome 1': [{'Intellectual developmental disorder, autosomal dominant'},
  {'coffin-siris syndrome 1'},
  {},
  {'Coffin-Siris syndrome'}],
 'bethlem myopathy 1': [{'bethlem myopathy 1'},
  {'Muscular dystrophy, limb-girdle, autosomal dominant'},
  {},
  {'Bethlem myopathy'}]}

In [11]:
descendants = {x: values for x in onto_src.getClasses() if len(values:=onto_src.getDescendants(x)) > 2}
descendant_names = {'\n'.join(onto_src.getPrefferedLabels(x)): 
            [onto_src.getPrefferedLabels(y) for y in values] for x, values in descendants.items()}
dict(list(descendant_names.items())[:5])

{'Aortic aneurysm, familial abdominal': [{'aortic aneurysm, familial abdominal, 2'},
  {'aortic aneurysm, familial abdominal, 1'},
  {'aortic aneurysm, familial abdominal, 3'},
  {'aortic aneurysm, familial abdominal, 4'},
  {'Aortic aneurysm, familial abdominal'}],
 'Adams-Oliver syndrome': [{'adams-oliver syndrome 2'},
  {'adams-oliver syndrome 1'},
  {'adams-oliver syndrome 6'},
  {'adams-oliver syndrome 4'},
  {'adams-oliver syndrome 3'},
  {'Adams-Oliver syndrome'},
  {'adams-oliver syndrome 5'}],
 'Acrodysostosis': [{'Acrodysostosis'},
  {'acrodysostosis 2 with or without hormone resistance'},
  {'acrodysostosis 1 with or without hormone resistance'}],
 'Pituitary adenoma': [{'Pituitary adenoma'},
  {'pituitary adenoma 2, growth hormone-secreting'},
  {'pituitary adenoma 4, acth-secreting'},
  {'pituitary adenoma 3, multiple types'},
  {'pituitary adenoma 5, multiple types'},
  {'pituitary adenoma 1, multiple types'}],
 'Restless legs syndrome': [{'restless legs syndrome, suscept

In [10]:
synonyms = {x:values for x in onto_src.getClasses() if len(values:=onto_src.getSynonymsNames(x))>0}
dict(list(synonyms.items())[:10])

{entry.100070: {'AAA1'},
 entry.100100: {'PBS'},
 entry.100300: {'AOS1'},
 entry.100800: {'ACH'},
 entry.101000: {'NF2'},
 entry.101400: {'SCS'},
 entry.101800: {'ACRDYS1'},
 entry.101850: {'PPKP3'},
 entry.101900: {'AKV'},
 entry.102200: {'PITA1'}}

In [11]:
annotation_labels = {x:values for x in onto_src.getClasses() if len(values:=onto_src.getAnnotationNames(x))>1}
dict(list(annotation_labels.items())[:10])

{entry.100070: {'AAA1', 'aortic aneurysm, familial abdominal, 1'},
 entry.100100: {'PBS', 'prune belly syndrome'},
 entry.100300: {'AOS1', 'adams-oliver syndrome 1'},
 entry.100800: {'ACH', 'achondroplasia'},
 entry.101000: {'NF2', 'neurofibromatosis,  iia 2'},
 entry.101400: {'SCS', 'saethre-chotzen syndrome'},
 entry.101800: {'ACRDYS1',
  'acrodysostosis 1 with or without hormone resistance'},
 entry.101850: {'PPKP3', 'palmoplantar keratoderma, punctate  iia 3'},
 entry.101900: {'AKV', 'acrokeratosis verruciformis'},
 entry.102200: {'PITA1', 'pituitary adenoma 1, multiple types'}}

In [12]:
annotation_labels = {x:values for x in onto_tgt.getClasses() if len(values:=onto_tgt.getAnnotationNames(x))>2}
dict(list(annotation_labels.items())[:10])

{ORDO.Orphanet_100006: {'ABeta amyloidosis, Dutch type',
  'ABetaE22Q amyloidosis',
  'HCHWA, Dutch type',
  'HCHWA-D',
  'Hereditary cerebral hemorrhage with amyloidosis, Dutch type'},
 ORDO.Orphanet_100008: {'ACys amyloidosis',
  'CST3-related amyloidosis',
  'Cystatin amyloidosis',
  'HCHWA, Icelandic type',
  'Hereditary cerebral hemorrhage with amyloidosis, Icelandic type',
  'Hereditary cystatin C amyloid angiopathy'},
 ORDO.Orphanet_100025: {'Alpha-HCD',
  'Alpha-heavy chain disease',
  'IPSID',
  'Immunoproliferative small intestinal disease',
  'Mediterranean lymphoma'},
 ORDO.Orphanet_100026: {'Franklin disease',
  'Gamma-HCD',
  'Gamma-heavy chain disease'},
 ORDO.Orphanet_100050: {'HAE 1',
  'HAE-I',
  'Hereditary angioedema type 1',
  'Hereditary angioneurotic edema type 1'},
 ORDO.Orphanet_528623: {'HAE with C1 inhibitor deficiency',
  'HAE with C1Inh deficiency',
  'Hereditary angioedema with C1Inh deficiency',
  'Hereditary angioneurotic edema with C1 inhibitor deficien

## Prompting

In [13]:
class OntologyEntryAttr:
    ## TODO: Add more attributes: direct children and parents
    def __init__(self, class_uri, onto: OntologyAccess, class_id=None):
        assert class_uri is not None or class_id is not None
        if class_uri is not None:
            self.class_id = onto.getClassByURI(class_uri)
        else:
            self.class_id = class_id
        assert self.class_id is not None
        self.annotation = {'class': self.class_id}
        self.onto:OntologyAccess = onto
        self.annotateEntry(onto)

    def annotateEntry(self, onto: OntologyAccess):
        logging.debug(f"Annotating {self.class_id}")
        self.annotation['children'] = onto.getDescendants(self.class_id)
        self.annotation['parents'] = onto.getAncestors(self.class_id)
        self.annotation['synonyms'] = onto.getSynonymsNames(self.class_id)
        self.annotation['names'] = onto.getAnnotationNames(self.class_id)

    def getAllEntetyNames(self):
        return self.annotation['names']

    def getSynonyms(self):
        return self.annotation['synonyms']
    
    def getChildren(self, skip_itself=True):
        result = self.annotation['children']
        if skip_itself:
            result = result.difference({self.class_id})
        return result

    def getParents(self, skip_itself=True):
        result = self.annotation['parents']
        if skip_itself:
            result = result.difference({self.class_id})
        return result

    def getParentsPreferredNames(self, skip_itself=True):
        labels = []
        for parent in self.getParents(skip_itself=skip_itself):
            preffered = self.onto.getPrefferedLabels(parent)
            if preffered:
                labels.append(preffered)
            else:
                labels.append({parent})
        return labels

    def getChildredPreferredNames(self, skip_itself=True):
        return [preffered if (preffered := self.onto.getPrefferedLabels(child)) else {child}
                for child in self.getChildren(skip_itself=skip_itself)]

    def __repr__(self):
        return str(self.annotation)

    def __str__(self):
        return str(self.annotation)

In [14]:
source_entity_uri = 'http://omim.org/entry/609560' #'http://omim.org/entry/100070'
target_entity_uri = 'http://www.orpha.net/ORDO/Orphanet_254875' #'http://omim.org/entry/101900'
source_entry = OntologyEntryAttr(source_entity_uri, onto_src)
target_entity = OntologyEntryAttr(target_entity_uri, onto_tgt)

In [15]:
display(source_entry)
print(target_entity)

{'class': entry.609560, 'children': {entry.609560}, 'parents': {phenotypicSeries.PS603041, owl.Thing, entry.609560}, 'synonyms': {'MTDPS2'}, 'names': {'MTDPS2', 'mitochondrial DNA depletion syndrome 2 (myopathic type)'}}

{'class': ORDO.Orphanet_254875, 'children': {ORDO.Orphanet_254875}, 'parents': {ORDO.Orphanet_254875, ORDO.Orphanet_C001, ORDO.Orphanet_377788, owl.Thing}, 'synonyms': {'mtDNA depletion syndrome, myopathic form'}, 'names': {'Mitochondrial DNA depletion syndrome, myopathic form', 'mtDNA depletion syndrome, myopathic form'}}


In [16]:
print(f" Source entity names: {source_entry.getAllEntetyNames()}")
print(f" Source entity synonyms: {source_entry.getSynonyms()}")
print(f" Source entity children names: {source_entry.getChildredPreferredNames(skip_itself=False)}")
print(f" Source entity children names: {source_entry.getChildredPreferredNames(skip_itself=True)}")
print(f" Source entity parents names: {source_entry.getParentsPreferredNames()}")
print(f" Source entity parents: {source_entry.getParents()}")

 Source entity names: {'MTDPS2', 'mitochondrial DNA depletion syndrome 2 (myopathic type)'}
 Source entity synonyms: {'MTDPS2'}
 Source entity children names: [{'mitochondrial DNA depletion syndrome 2 (myopathic type)'}]
 Source entity children names: []
 Source entity parents names: [{owl.Thing}, {'Mitochondrial DNA depletion syndrome'}]
 Source entity parents: {owl.Thing, phenotypicSeries.PS603041}


In [19]:
def composePrompt(source_entry:OntologyEntryAttr, target_entry:OntologyEntryAttr):
    return f"""
**Task Description:**
Given two sets of concepts with their names, parent relationships, and child relationships, determine if these concepts are identical. Evaluate based on the provided names and hierarchical structures.
**Template for Alignment:**

1. **Source Concept:**
**Names:** [List of source concept names]
**Parent Concepts:** [List of source concept's parent names]
**Child Concepts:** [List of source concept's child names]
2. **Target Concept:**
**Names:** [List of target concept names]
**Parent Concepts:** [List of target concept's parent names]
**Child Concepts:** [List of target concept's child names]
**Determine Concept Equivalence:**
Analyze whether the source and target concepts are equivalent based on names, parent concepts, and child concepts. Provide your conclusion ("Yes" for identical, "No" for different).

**Example Prompts (Few-Shot Examples):**

1. **Example 1:**
Source Concept Names: [e.g., "Entity_A"]
Parent of Source Concept: [e.g., "Parent_A"]
Child of Source Concept: [e.g., "Child_A"]
Target Concept Names: [e.g., "Entity_B"]
Parent of Target Concept: [e.g., "Parent_B"]
Child of Target Concept: [e.g., "Child_B"]

**Conclusion:** No

2. **Example 2:**
Source Concept Names: [e.g., "Entity_X"]
Parent of Source Concept: [e.g., "Parent_X"]
Child of Source Concept: [e.g., "Child_X"]
Target Concept Names: [e.g., "Entity_X"]
Parent of Target Concept: [e.g., "Parent_X"]
Child of Target Concept: [e.g., "Child_X"]
**Conclusion:** Yes
**Now Answer:**

Source Concept Names: {source_entry.getAllEntetyNames()}
Parent of Source Concept: {source_entry.getParentsPreferredNames()}
Child of Source Concept: {source_entry.getChildredPreferredNames()}
Target Concept Names: {target_entry.getAllEntetyNames()}
Parent of Target Concept: {target_entry.getParentsPreferredNames()}
Child of Target Concept: {target_entry.getChildredPreferredNames()}
""" 

In [20]:
source_entity_uri = 'http://omim.org/entry/609560' #'http://omim.org/entry/100070'
target_entity_uri = 'http://www.orpha.net/ORDO/Orphanet_254875' #'http://omim.org/entry/101900'
source_entity = OntologyEntryAttr(source_entity_uri, onto_src)
target_entity = OntologyEntryAttr(target_entity_uri, onto_tgt)
demo_prompt = composePrompt(source_entity, target_entity)
print(demo_prompt)


**Task Description:**
Given two sets of concepts with their names, parent relationships, and child relationships, determine if these concepts are identical. Evaluate based on the provided names and hierarchical structures.
**Template for Alignment:**

1. **Source Concept:**
**Names:** [List of source concept names]
**Parent Concepts:** [List of source concept's parent names]
**Child Concepts:** [List of source concept's child names]
2. **Target Concept:**
**Names:** [List of target concept names]
**Parent Concepts:** [List of target concept's parent names]
**Child Concepts:** [List of target concept's child names]
**Determine Concept Equivalence:**
Analyze whether the source and target concepts are equivalent based on names, parent concepts, and child concepts. Provide your conclusion ("Yes" for identical, "No" for different).

**Example Prompts (Few-Shot Examples):**

1. **Example 1:**
Source Concept Names: [e.g., "Entity_A"]
Parent of Source Concept: [e.g., "Parent_A"]
Child of Sour

In [21]:
response = mistral.ask_question(demo_prompt)
print(response)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.


**Conclusion:** No

The source concept and target concept have different names, even though they seem to refer to the same entity (Mitochondrial DNA depletion syndrome, myopathic type). However, the hierarchical structures are not identical. The source concept does not have any parent concepts specified, while the target concept has multiple parent concepts, including 'disease' and 'clinical entity'. Therefore, they are not equivalent based on the provided information.


## Local matching

In [17]:
import pandas as pd
test_candidates_path = DATASET_PATH / "omim-ordo/refs_equiv/test.cands.tsv"
test_candidates = pd.read_csv(test_candidates_path, sep='\t')
test_candidates['TgtCandidates'] = test_candidates['TgtCandidates'].str.strip("('").str.strip("')").str.split("', '")

In [141]:
test_candidates.head()

SrcEntity  \
0  http://www.omim.org/phenotypicSeries/PS309510   
1                   http://omim.org/entry/615075   
2                   http://omim.org/entry/613724   
3                   http://omim.org/entry/602066   
4                   http://omim.org/entry/167730   

                                   TgtEntity  \
0   http://www.orpha.net/ORDO/Orphanet_98464   
1  http://www.orpha.net/ORDO/Orphanet_404473   
2  http://www.orpha.net/ORDO/Orphanet_163684   
3   http://www.orpha.net/ORDO/Orphanet_31709   
4    http://www.orpha.net/ORDO/Orphanet_2399   

                                       TgtCandidates  
0  [http://www.orpha.net/ORDO/Orphanet_93406, htt...  
1  [http://www.orpha.net/ORDO/Orphanet_157791, ht...  
2  [http://www.orpha.net/ORDO/Orphanet_639, http:...  
3  [http://www.orpha.net/ORDO/Orphanet_99141, htt...  
4  [http://www.orpha.net/ORDO/Orphanet_1475, http...

In [155]:
# src_onto_path = DATASET_PATH / 'omim-ordo/omim.owl'
# tgt_onto_path = DATASET_PATH / 'omim-ordo/ordo.owl'

# onto_src = OntologyAccess(src_onto_path, annotate_on_init=True)
# onto_tgt = OntologyAccess(tgt_onto_path, annotate_on_init=True)

In [ ]:
def dummy_prompt(src_entety:OntologyEntryAttr, tgt_entety:OntologyEntryAttr):
    return f"""
    **Task Description:**
    Given two entities from different ontologies with their names, parent relationships, and child relationships, determine if these concepts are the same:

    1. **Source Entity:**
    **All Entity names:** {src_entety.getAllEntetyNames()}
    **Parent Entity Namings:** {src_entety.getParentsPreferredNames()}
    **Child Entity Namings:** {src_entety.getChildredPreferredNames()}

    2. **Target Entity:**
    **All Entity names:** {tgt_entety.getAllEntetyNames()}
    **Parent Entity Namings:** {tgt_entety.getParentsPreferredNames()}
    **Child Entity Namings:** {tgt_entety.getChildredPreferredNames()}

    Write "Yes" if the entities refer to the same concepts, and "No" otherwise.
    """

def dummy_answer_extraction(response:str):
    """
    Returns 'Yes', 'No' or None based on the response, 
    Decision is made based on the presence of positive or negative words in the response
    """
    response = response.lower().strip()
    positive_words = ["yes", "yeah", "yep", "correct", "true"]
    negative_words = ["no", "nah", "nope", "wrong", "false"]
    is_positive = any([word in response for word in positive_words])
    is_negative = any([word in response for word in negative_words])
    if is_positive and not is_negative:
        return "Yes"
    elif is_negative and not is_positive:
        return "No"
    elif not is_positive and not is_negative:
        return None
    response = response.split('\n')
    for line in response:
        if "yes" in line:
            return "Yes"
        elif "no" in line:
            return "No"
    return None
    

In [18]:
test_case = test_candidates.iloc[0]

test_case_results_df = pd.DataFrame(columns=['SrcEntity', 'TgtCandidate', 'ModelAnswer', 'RealAnswer'])
test_case_results_df['TgtCandidate'] = test_case['TgtCandidates']
test_case_results_df['SrcEntity'] = test_case['SrcEntity']
test_case_results_df['RealAnswer'] = test_case_results_df['TgtCandidate'].apply(lambda x: "Yes" if x == test_case['TgtEntity'] else "No")
test_case_results_df

SrcEntity  \
0    http://www.omim.org/phenotypicSeries/PS309510   
1    http://www.omim.org/phenotypicSeries/PS309510   
2    http://www.omim.org/phenotypicSeries/PS309510   
3    http://www.omim.org/phenotypicSeries/PS309510   
4    http://www.omim.org/phenotypicSeries/PS309510   
..                                             ...   
96   http://www.omim.org/phenotypicSeries/PS309510   
97   http://www.omim.org/phenotypicSeries/PS309510   
98   http://www.omim.org/phenotypicSeries/PS309510   
99   http://www.omim.org/phenotypicSeries/PS309510   
100  http://www.omim.org/phenotypicSeries/PS309510   

                                  TgtCandidate ModelAnswer RealAnswer  
0     http://www.orpha.net/ORDO/Orphanet_93406         NaN         No  
1       http://www.orpha.net/ORDO/Orphanet_181         NaN         No  
2      http://www.orpha.net/ORDO/Orphanet_3231         NaN         No  
3    http://www.orpha.net/ORDO/Orphanet_293642         NaN         No  
4     http://www.orpha.net/ORDO/Orphanet_90025         NaN         No  
..                                         ...         ...        ...  
96   http://www.orpha.net/ORDO/Orphanet_163209         NaN         No  
97   http://www.orpha.net/ORDO/Orphanet_225147         NaN         No  
98   http://www.orpha.net/ORDO/Orphanet_466084         NaN         No  
99   http://www.orpha.net/ORDO/Orphanet_183530         NaN         No  
100   http://www.orpha.net/ORDO/Orphanet_85275         NaN         No  

[101 rows x 4 columns]

In [ ]:
src_entity = OntologyEntryAttr(test_case['SrcEntity'], onto_src)
tgt_entity = OntologyEntryAttr(test_case['TgtEntity'], onto_tgt)

candidate_idx = 0
candidate_entity = OntologyEntryAttr(test_case['TgtCandidates'][candidate_idx], onto_tgt)

prompt = dummy_prompt(src_entity, candidate_entity)
# model_response = mistral.ask_question(prompt)
# pred_answer = dummy_answer_extraction(model_response)
print(prompt)


    **Task Description:**
    Given two sets of concepts with their names, parent relationships, and child relationships, determine if these concepts are the same:

    1. **Source Concept:**
    **All its names:** {'Intellectual developmental disorder, X-linked syndromic'}
    **Parent Concept Namings:** [{owl.Thing}]
    **Child Concept Namings:** [{'intellectual developmental disorder, x-linked, syndromic 11'}, {'prieto X-linked mental retardation syndrome'}, {'basilicata-akhtar syndrome'}, {'mehmo syndrome'}, {'abidi X-linked mental retardation syndrome'}, {'corpus callosum, agenesis of, with impaired intellectual development, ocular coloboma, and micrognathia'}, {"van esch-o'driscoll syndrome"}, {'raynaud-claes syndrome'}, {'arts syndrome'}, {'intellectual developmental disorder, x-linked, syndromic, lubs  iia'}, {'intellectual developmental disorder, x-linked, syndromic 7'}, {'intellectual developmental disorder, x-linked, syndromic, siderius  iia'}, {'paganini-miozzo syndrome'}

In [ ]:
import random
# set logging level to ignore most of the logs
logging.getLogger().setLevel(logging.WARNING)

src_entity = OntologyEntryAttr(test_case['SrcEntity'], onto_src)
for idx, row in test_case_results_df.iterrows():
    candidate_entity = OntologyEntryAttr(row['TgtCandidate'], onto_tgt)
    prompt = dummy_prompt(src_entity, candidate_entity)
    # model_response = mistral.ask_question(prompt)
    model_response = 'Yes.' if random.random() > 0.8 else 'No.'
    pred_answer = dummy_answer_extraction(model_response)
    test_case_results_df.loc[idx, 'ModelResponse'] = model_response
    test_case_results_df.loc[idx, 'ModelAnswer'] = pred_answer
    # print(prompt)

In [199]:
positive_targets = test_case_results_df[test_case_results_df['ModelAnswer'] == 'Yes']
print(f"Number of positively labeled targets: {len(positive_targets)}")
positive_targets

Number of positively labeled targets: 11


SrcEntity  \
0   http://www.omim.org/phenotypicSeries/PS309510   
9   http://www.omim.org/phenotypicSeries/PS309510   
24  http://www.omim.org/phenotypicSeries/PS309510   
25  http://www.omim.org/phenotypicSeries/PS309510   
70  http://www.omim.org/phenotypicSeries/PS309510   
74  http://www.omim.org/phenotypicSeries/PS309510   
75  http://www.omim.org/phenotypicSeries/PS309510   
83  http://www.omim.org/phenotypicSeries/PS309510   
86  http://www.omim.org/phenotypicSeries/PS309510   
92  http://www.omim.org/phenotypicSeries/PS309510   
97  http://www.omim.org/phenotypicSeries/PS309510   

                                 TgtCandidate ModelAnswer RealAnswer  
0    http://www.orpha.net/ORDO/Orphanet_93406         Yes         No  
9    http://www.orpha.net/ORDO/Orphanet_48471         Yes         No  
24   http://www.orpha.net/ORDO/Orphanet_85274         Yes         No  
25  http://www.orpha.net/ORDO/Orphanet_295195         Yes         No  
70    http://www.orpha.net/ORDO/Orphanet_1576         Yes         No  
74  http://www.orpha.net/ORDO/Orphanet_441434         Yes         No  
75   http://www.orpha.net/ORDO/Orphanet_93593         Yes         No  
83   http://www.orpha.net/ORDO/Orphanet_68385         Yes         No  
86   http://www.orpha.net/ORDO/Orphanet_98565         Yes         No  
92  http://www.orpha.net/ORDO/Orphanet_163971         Yes         No  
97  http://www.orpha.net/ORDO/Orphanet_225147         Yes         No

### Extencive matching

In [72]:
def implication_prompt(source_entry, target_entry):
    """
    Creates a prompt asking if 'source_entry' implies 'target_entry'.
    Example: "It is [a/an] C, is it a D? Yes/No"
    """
    return f"It is [a/an] ${source_entry}$, is it a ${target_entry}$? Yes/No"

def equivalence_prompt(source_entry, target_entry):
    """
    Creates prompts for both directions to check if 'source_entry' implies 'target_entry' and vice versa.
    - Example: "It is [a/an] C, is it a D? Yes/No"
    - Example: "It is [a/an] D, is it a C? Yes/No"
    """
    prompt1 = implication_prompt(source_entry, target_entry)
    prompt2 = implication_prompt(target_entry, source_entry)
    return prompt1, prompt2

def synonyms_prompt(source_label, target_label):
    """
    Uses ontology to create prompts involving synonyms for C and D.
    Example: "It is [a/an] C-label2, is it a D-label3? Yes/No"
    """
    return f"It is [a/an] ${source_label}$, is it a ${target_label}$? Yes/No"

def transitivity_prompt(source_entry, intermediate_entry, target_entry):
    """
    Creates a transitivity prompt to check if 'source_entry' implies an intermediate entry,
    which then implies 'target_entry'.
    """
    return f"It is [a/an] ${source_entry}$, implies ${intermediate_entry}$, implies ${target_entry}$? Yes/No"

def hierarchical_equivalence_prompt(child1, parent1, child2, parent2):
    """
    Creates prompts to check bidirectional equivalence between two hierarchical entries.
    Example:
    - "Something is [a/an] C and [a/an] C', is it also [a/an] D and [a/an] D'? Yes/No"
    - "Something is [a/an] D and [a/an] D', is it also [a/an] C and [a/an] C'? Yes/No"
    """
    prompt1 = f"Something is [a/an] ${child1}$ and [a/an] ${parent1}$, is it also [a/an] ${child2}$ and [a/an] ${parent2}$? Yes/No"
    prompt2 = f"Something is [a/an] ${child2}$ and [a/an] ${parent2}$, is it also [a/an] ${child1}$ and [a/an] ${parent1}$? Yes/No"
    return prompt1, prompt2

In [ ]:
def local_matching_confidence(source_entry, target_entry, onto_1=None, onto_2=None):
    # TODO: change this function to use the General Model
    # TODO: implement response parsing
    source_attr = OntologyEntryAttr(source_entry, onto_1)
    target_attr = OntologyEntryAttr(target_entry, onto_2)

    general_points = 0
    retrieved_points = 0

    # Note: All names already includes Synonyms and Main Labels as well
    # compare names
    for s in source_attr.getAllEntetyNames():
        for t in target_attr.getAllEntetyNames():
            e1, e2 = equivalence_prompt(s, t)
            q1 = mistral.ask_question(e1)
            q2 = mistral.ask_question(e2)
            if  "Yes" in q1:
                retrieved_points += 1
            if "Yes" in q2:
                retrieved_points += 1
            if "Yes" in q1 and "Yes" in q2:
                retrieved_points += 1
            print(retrieved_points/(general_points + 1e-6))
            general_points += 3

    # compare synonyms
    for s in source_attr.getSynonyms():
        for t in target_attr.getSynonyms():
            e = synonyms_prompt(s, t)
            q = mistral.ask_question(e)
            if "Yes" in q:
                retrieved_points += 1
            general_points += 1
            print(retrieved_points/(general_points + 1e-6))

    # compare hierarchy
    source_hierarchy = {
        "children": source_attr.getChildren(),
        "parents": source_attr.getParents()
    }

    target_hierarchy = {
        "children": target_attr.getChildren(),
        "parents": target_attr.getParents()
    }

    for schild in source_hierarchy["children"]:
        for tchild in target_hierarchy["children"]:
            if source_hierarchy["parents"] and target_hierarchy["parents"]:
                for sparent, tparent in zip(source_hierarchy["parents"], target_hierarchy["parents"]):
                    e1, e2 = hierarchical_equivalence_prompt(schild, sparent, tchild, tparent)
                    
                    q1 = mistral.ask_question(e1)
                    q2 = mistral.ask_question(e2)
                    
                    retrieved_points += ("Yes" in q1) + ("Yes" in q2)
                    if "Yes" in q1 and "Yes" in q2:
                        retrieved_points += 1
                    
                    general_points += 3
                    print(retrieved_points / (general_points + 1e-6))


    confidence = retrieved_points / general_points
    return confidence

In [ ]:
!pip install tqdm -q

In [ ]:
# Uncomment to calculate confidences

# from tqdm import tqdm

# confidences = {}

# for _, row in tqdm(test_candidates.iterrows(), total=test_candidates.shape[0], desc="Processing rows"):
#     source = row[0]
#     target = row[1]
#     confidences[(source, target)] = []

#     for candidate in tqdm(row[2], desc=f"Processing candidates for ({source}, {target})", leave=False):
#         confidence = local_matching_confidence(source, candidate, onto_1=onto_src, onto_2=onto_tgt)
#         confidences[(source, target)].append((candidate, confidence))

# Initial Filtering with Embedding Model

In [19]:
!pip install sentence_transformers -q

In [20]:
import torch
import re
from sentence_transformers import SentenceTransformer

In [21]:
model = SentenceTransformer("dunzhang/stella_en_1.5B_v5", trust_remote_code=True).cuda()

modules.json:   0%|          | 0.00/316 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/397 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/174k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/51.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/844 [00:00<?, ?B/s]

modeling_qwen.py:   0%|          | 0.00/65.3k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/dunzhang/stella_en_1.5B_v5:
- modeling_qwen.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


model.safetensors:   0%|          | 0.00/6.17G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.31k [00:00<?, ?B/s]

tokenization_qwen.py:   0%|          | 0.00/10.8k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/dunzhang/stella_en_1.5B_v5:
- tokenization_qwen.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


vocab.json:   0%|          | 0.00/2.78M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.67M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/7.03M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/80.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/370 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/289 [00:00<?, ?B/s]

2_Dense_1024/config.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/6.30M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/6.30M [00:00<?, ?B/s]

In [23]:
similarity_threshold = 0.9
similarities = {}

In [36]:
src_entity_name = list(OntologyEntryAttr(test_case['SrcEntity'], onto_src).getAllEntetyNames())[0]
tgt_entity_name = list(OntologyEntryAttr(test_case['TgtEntity'], onto_tgt).getAllEntetyNames())[0]

In [37]:
for _, row in test_case_results_df.iterrows():
    tgt_candidate = row['TgtCandidate']
    candidate_entity_name = list(OntologyEntryAttr(tgt_candidate, onto_tgt).getAllEntetyNames())[0]
    similarity = model.similarity(model.encode(src_entity_name), model.encode(candidate_entity_name))
    similarities[tgt_candidate] = similarity
    print(similarity)

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5686]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7394]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6001]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.4531]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5845]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6920]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6429]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5378]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5914]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6892]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7614]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6380]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6298]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8480]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8968]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5829]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8056]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.4485]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6106]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8896]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6803]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6057]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6267]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7821]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.4586]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.4278]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6732]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6979]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6718]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7708]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8184]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6886]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6201]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5888]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7738]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.9064]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7039]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6095]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6803]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6253]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6481]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6833]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8049]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7803]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7483]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5454]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7138]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7039]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5976]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8417]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8904]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6798]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5479]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6552]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6028]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.9619]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5758]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5562]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8513]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5651]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5665]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7437]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6188]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.9385]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.4956]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6115]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5636]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7286]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.4649]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7317]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7040]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.3911]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6162]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7363]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6594]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.4695]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6207]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8416]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7169]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6692]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7883]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6471]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7754]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6430]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8601]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.4947]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6326]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6052]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6500]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6231]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6777]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8518]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.8965]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6078]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6172]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.7057]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6233]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6310]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.5967]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6867]])


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

tensor([[0.6242]])


In [39]:
print(similarities)

{'http://www.orpha.net/ORDO/Orphanet_93406': tensor([[0.5686]]), 'http://www.orpha.net/ORDO/Orphanet_181': tensor([[0.7394]]), 'http://www.orpha.net/ORDO/Orphanet_3231': tensor([[0.6001]]), 'http://www.orpha.net/ORDO/Orphanet_293642': tensor([[0.4531]]), 'http://www.orpha.net/ORDO/Orphanet_90025': tensor([[0.5845]]), 'http://www.orpha.net/ORDO/Orphanet_90625': tensor([[0.6920]]), 'http://www.orpha.net/ORDO/Orphanet_324761': tensor([[0.6429]]), 'http://www.orpha.net/ORDO/Orphanet_596': tensor([[0.5378]]), 'http://www.orpha.net/ORDO/Orphanet_35093': tensor([[0.5914]]), 'http://www.orpha.net/ORDO/Orphanet_48471': tensor([[0.6892]]), 'http://www.orpha.net/ORDO/Orphanet_423655': tensor([[0.7614]]), 'http://www.orpha.net/ORDO/Orphanet_96265': tensor([[0.6380]]), 'http://www.orpha.net/ORDO/Orphanet_199633': tensor([[0.6298]]), 'http://www.orpha.net/ORDO/Orphanet_88616': tensor([[0.8480]]), 'http://www.orpha.net/ORDO/Orphanet_182076': tensor([[0.8968]]), 'http://www.orpha.net/ORDO/Orphanet_371

In [41]:
import csv

with open('output.csv', 'w', newline='') as csv_file:
    writer = csv.writer(csv_file)
    writer.writerow(['Key', 'Value'])  # Header row
    for key, value in similarities.items():
        writer.writerow([key, value.item()])

print("Data saved to output.csv")

Data saved to output.csv


# Prev evaluation sample

In [ ]:
from tqdm import tqdm

ranked_candidates = {}

for source, target, candidates in tqdm(test_candidates.iterrows(), total=test_candidates.shape[0], desc="Processing rows"):
    ranked_candidates[(source, target)] = sorted(
        [(candidate, local_matching_confidence(source, candidate)) 
         for candidate in tqdm(candidates, desc=f"Processing candidates for {source}", leave=False)],
        key=lambda x: x[1],
        reverse=True
    )

In [ ]:
def hits_at_K(test_data, ranked_candidates, K):
    hits = 0
    for source, target, _, in test_candidates.iterrows():
        top_k_candidates = [candidate for candidate, _ in ranked_candidates[(source, target)][:K]]
        if target in top_k_candidates:
            hits += 1
    return hits / len(test_data)

In [ ]:
K = 5
hits = hits_at_K(test_candidates, ranked_candidates, K)
print(f"Hits@{K}: {hits:.2f}")

# Testing

### Test how parents influence KG alignment
Hypothesis: Including parent information in prompts improves alignment accuracy.

In [ ]:
def test_information_effect_parents(source_entry: OntologyEntryAttr, target_entry: OntologyEntryAttr):
    return f"""
        **Task Description:**
        Determine if the following entities from two different ontologies are equivalent. Consider their names and parent relationships:

        1. **Source Entity:**
        **All Entity Names:** {source_entry.getAllEntetyNames()}
        **Parent Names:** {source_entry.getParentsPreferredNames()}

        2. **Target Entity:**
        **All Entity Names:** {target_entry.getAllEntetyNames()}
        **Parent Names:** {target_entry.getParentsPreferredNames()}

        Write "Yes" if the entities refer to the same concept, and "No" otherwise.
        """

In [ ]:
def test_information_effect_no_parents(source_entry: OntologyEntryAttr, target_entry: OntologyEntryAttr):
    return f"""
        **Task Description:**
        Determine if the following entities from two different ontologies are equivalent. Only consider their names:

        1. **Source Entity:**
        **All Entity Names:** {source_entry.getAllEntetyNames()}

        2. **Target Entity:**
        **All Entity Names:** {target_entry.getAllEntetyNames()}

        Write "Yes" if the entities refer to the same concept, and "No" otherwise.
        """

### Testing Hierarchy Awareness

Hypothesis: Using hierarchical relationships (e.g., parent-child) improves alignment precision

In [ ]:
def test_full_hierarchy_effect(source_entry: OntologyEntryAttr, target_entry: OntologyEntryAttr):
    return f"""
    **Task Description:**
    Evaluate whether the source and target entities represent the same concept, considering hierarchical relationships:

    1. **Source Entity:**
    **All Entity Names:** {source_entry.getAllEntetyNames()}
    **Parent Names:** {source_entry.getParentsPreferredNames()}
    **Child Names:** {source_entry.getChildredPreferredNames()}

    2. **Target Entity:**
    **All Entity Names:** {target_entry.getAllEntetyNames()}
    **Parent Names:** {target_entry.getParentsPreferredNames()}
    **Child Names:** {target_entry.getChildredPreferredNames()}

    Write "Yes" if the entities refer to the same concept, and "No" otherwise.
    """

### Synonym Comparison Prompt

This prompt directly compares synonyms of the source and target entities.

In [ ]:
def compare_synonyms_prompt(source_entry: OntologyEntryAttr, target_entry: OntologyEntryAttr):
    """
    Creates a prompt asking if the synonyms of the source and target entries indicate equivalence.
    """
    return f"""
    **Task Description:**
    Given two entities with their synonyms from different ontologies, determine if they refer to the same concept.

    1. **Source Entity:**
    **Synonyms:** {source_entry.getSynonyms()}

    2. **Target Entity:**
    **Synonyms:** {target_entry.getSynonyms()}

    Write "Yes" if the entities refer to the same concept, and "No" otherwise.
    """

**Synonym Inclusion with Parent-Child Context**


Combines synonym information with hierarchical relationships for richer comparisons.

In [ ]:
def synonyms_with_hierarchy_prompt(source_entry, target_entry):
    """
    Creates a prompt that includes synonyms along with parent-child relationships.
    """
    return f"""
    **Task Description:**
    Given two entities from different ontologies with their synonyms, parent, and child relationships, determine if they refer to the same concept.

    1. **Source Entity:**
    **Synonyms:** {source_entry.getSynonyms()}
    **Parent Entity Namings:** {source_entry.getParentsPreferredNames()}
    **Child Entity Namings:** {source_entry.getChildredPreferredNames()}

    2. **Target Entity:**
    **Synonyms:** {target_entry.getSynonyms()}
    **Parent Entity Namings:** {target_entry.getParentsPreferredNames()}
    **Child Entity Namings:** {target_entry.getChildredPreferredNames()}

    Write "Yes" if the entities refer to the same concept, and "No" otherwise.
    """

**Synonym-Based Subsumption Prompt** ???

Checks if the synonyms indicate a subsumption relationship between the entities.

In [ ]:
def synonyms_subsumption_prompt(source_entry, target_entry):
    """
    Creates a prompt to determine if one entity's synonyms suggest it subsumes the other.
    """
    return f"""
    **Task Description:**
    Given two entities with their synonyms, determine if one entity subsumes the other.

    1. **Source Entity Synonyms:** {source_entry.getSynonyms()}
    2. **Target Entity Synonyms:** {target_entry.getSynonyms()}

    Write "Yes" if the entities refer to the same concept, and "No" otherwise.
    """

**Synonym Disjointness Hypothesis Prompt** ???

Tests if synonyms indicate disjointness between the entities.

In [ ]:
def synonym_disjointness_prompt(source_entry, target_entry):
    """
    Creates a prompt to hypothesize if the two entities are disjoint based on their synonyms.
    """
    return f"""
    **Task Description:**
    Using the synonyms of two entities, determine if they refer to disjoint concepts.

    1. **Source Entity Synonyms:** {source_entry.getSynonyms()}
    2. **Target Entity Synonyms:** {target_entry.getSynonyms()}

    Write "Yes" if the entities refer to the same concept, and "No" otherwise.
    """

Compare prompt results

In [ ]:
prompt_functions = [
    test_information_effect_parents,
    test_information_effect_no_parents
    test_full_hierarchy_effect,
    compare_synonyms_prompt,
    synonyms_with_hierarchy_prompt,
    synonyms_subsumption_prompt,
    synonym_disjointness_prompt
]

models = {
    "mistral": mistral,
    # "open_ai": open_ai,
    # Add more models if needed
}

In [ ]:
def compose_prompts(source_entry: OntologyEntryAttr, target_entry: OntologyEntryAttr):
    composed_prompts = []
    for func in prompt_functions:
        prompt = func(source_entry, target_entry)
        composed_prompts.append(prompt)

    return composed_prompts

In [ ]:
import pandas as pd

def generate_results_table(test_case, models):
    results = []

    source_entry = test_case['SrcEntity']
    target_candidates = test_case['TgtCandidates']
    real_answer = test_case['TgtEntity']

    for candidate in target_candidates:
        prompts = compose_prompts(source_entry, candidate)

        for idx, prompt in enumerate(prompts):
            row = {
                'SrcEntity': source_entry,
                'TgtCandidate': candidate,
                'PromptNum': idx,
                'RealAnswer': "Yes" if candidate == real_answer else "No"
            }
            
            # Collect model answers for the current prompt
            for model_name, model_func in models.items():
                row[model_name] = model_func(prompt)
            
            results.append(row)

    results_df = pd.DataFrame(results)

    column_order = ['SrcEntity', 'TgtCandidate', 'Prompt', 'RealAnswer'] + list(models.keys())
    results_df = results_df[column_order]

    return results_df

In [ ]:
# test_case = test_candidates.iloc[0] ???
test_case = {
    "SrcEntity": "Entity_A",
    "TgtCandidates": ["Entity_B", "Entity_C"],
    "TgtEntity": "Entity_B"
}

results_table = generate_results_table(test_case, models, prompt_functions)
results_table

**Output should look like this**

<table border="1">
  <thead>
    <tr>
      <th>SrcEntity</th>
      <th>TgtCandidate</th>
      <th>PromptIdx</th>
      <th>RealAnswer</th>
      <th>Mistral</th>
      <th>OpenAI</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Entity_A</td>
      <td>Entity_B</td>
      <td>1</td>
      <td>Yes</td>
      <td>Yes</td>
      <td>No</td>
    </tr>
    <tr>
      <td>Entity_A</td>
      <td>Entity_B</td>
      <td>2</td>
      <td>Yes</td>
      <td>No</td>
      <td>Yes</td>
    </tr>
    <tr>
      <td>Entity_A</td>
      <td>Entity_C</td>
      <td>1</td>
      <td>No</td>
      <td>No</td>
      <td>Yes</td>
    </tr>
    <tr>
      <td>Entity_A</td>
      <td>Entity_C</td>
      <td>2</td>
      <td>No</td>
      <td>Yes</td>
      <td>No</td>
    </tr>
  </tbody>
</table>

# Evaluation

### Model Accuracy

Definition: Measure how well each model’s predictions align with the RealAnswer.
- Evaluation:
- Compute the overall accuracy for each model:

$\text{Accuracy} = \frac{\text{Number of Correct Predictions}}{\text{Total Predictions}}$

Example output:
- Mistral Accuracy: 80%
- OpenAI Accuracy: 75%


### Prompt(Hierarchy) Effectiveness

- Definition: Evaluate which prompts are most effective in helping models produce correct answers.
- Evaluation:
- Group results by Prompt and calculate the accuracy for each prompt across all models.
- Identify prompts with the highest accuracy.
- If certain prompts consistently lead to incorrect predictions, these can be flagged for revision.

Example Output:
- Prompt 1 Accuracy: 85%
- Prompt 2 Accuracy: 70%

### Model-Prompt Interaction

- Definition: Assess which prompts work best for specific models.
- Evaluation:
- Calculate per-prompt accuracy for each model.
- Identify combinations where certain models perform exceptionally well or poorly with specific prompts.

Example Output:

Prompt 1:
- Mistral: 90%
- OpenAI: 80%

### Error Analysis ???

- Definition: Identify patterns in errors made by the models.
- Evaluation:
- Filter rows where ModelAnswer does not match RealAnswer.
- Group errors by TgtCandidate, Prompt, or Model to identify common failure cases.
- Examine the characteristics of frequent errors (e.g., ambiguous prompts, challenging candidates).

Example Insights:

- “Most errors occur when the target candidate is very similar to the source entity.”
- “Model A struggles with Prompt 2, likely due to unclear phrasing.”

### Comparison Between Models

- Definition: Highlight strengths and weaknesses of each model relative to others.
- Evaluation:
- Compare accuracy, precision, recall, and F1 scores for each model.
- Identify cases where one model consistently outperforms the other, and hypothesize reasons (e.g., prompt comprehension, response generation).

Example Output:

Mistral:
  - Accuracy: 85%
  - F1 Score: 82%

OpenAI:
  - Accuracy: 75%
  - F1 Score: 77%

### Aggregated Results

- Definition: Summarize overall results to draw broad conclusions.
- Evaluation:
- Present a table or chart showing accuracy, precision, and recall for all models and prompts.

<table border="1">
  <tr>
    <th>Model</th>
    <th>Overall Accuracy</th>
    <th>Best Prompt Accuracy</th>
    <th>Worst Prompt Accuracy</th>
  </tr>
  <tr>
    <td>Mistral</td>
    <td>85%</td>
    <td>90%</td>
    <td>70%</td>
  </tr>
  <tr>
    <td>OpenAI</td>
    <td>75%</td>
    <td>80%</td>
    <td>60%</td>
  </tr>
</table>


### Statistical Testing
- Definition: Conduct statistical tests to determine if differences between models or prompts are significant.

Evaluation:
- Use metrics like p-values to confirm whether observed differences in accuracy are statistically significant.
- Identify whether one model is objectively better.


### Visualizations for the Report

Graphs and Charts:
- Bar Chart: Accuracy of each model.
- Heatmap: Model accuracy per prompt.
- Confusion Matrix: Display true positives, false positives, etc.
- Line Chart: Model performance over different test cases.